### PR TITLE
Add experience

### PR DESCRIPTION
**Pull request by Arcoz0308**
this PR add experiencemanager and function to update experience,
maybe i add xp orbs latter

### CODE REVIEW

File: server/entity/experience.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+var maxLevel = 21474836477
+
+type ExperienceManager struct {
+	level    int
+	progress float64
+	totalXP  int
+}
+
+func NewExperienceManager() *ExperienceManager {
+	return &ExperienceManager{
```
Comment (TwistedAsylumMC) [inline server/entity/experience.go:14]: You can do just ``&ExperienceManager{}``

File: server/entity/experience.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
+package entity
+
+var maxLevel = 21474836477
```
Comment (TwistedAsylumMC) [inline server/entity/experience.go]: Use math.MaxInt32 instead to prevent hard coding numbers

File: server/entity/experience.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+func (e *ExperienceManager) MaxLevel() int {
+	return maxLevel
+}
+func (e *ExperienceManager) TotalXP() int {
+	return e.totalXP
+}
+
+// SetTotalXP internal
+func (e *ExperienceManager) SetTotalXP(amount int) {
+	e.totalXP = amount
```
Comment (TwistedAsylumMC) [inline server/entity/experience.go:35]: Do you not need to recalculate the progress?

File: server/session/player.go [hunk 4]
```diff
@@ -605,6 +605,23 @@ func protocolToSkin(sk protocol.Skin) (s skin.Skin, err error) {
 	}
 	return
 }
+func (s *Session) SendXpValue(e *entity.ExperienceManager) {
+	s.writePacket(&packet.UpdateAttributes{
```
Comment (TwistedAsylumMC) [inline server/session/player.go:609]: Use XP instead of Xp

**Review by TwistedAsylumMC — CHANGES_REQUESTED** [review]
The project needs to be formatted as well as documentation for most methods

File: server/entity/experience.go (hunk 3 above)
Comment (Arcoz0308) [reply server/entity/experience.go]: i have remove useless function

File: server/entity/experience.go (hunk 1 above)
Comment (Arcoz0308) [reply server/entity/experience.go]: fix

File: server/entity/experience.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	return maxLevel
+}
+
+// TotalXP get the total xp collected.
+func (e *ExperienceManager) TotalXP() int {
+	return e.totalXP
+}
+
+// GetExperienceToNextLevel get the total experience netted to level up.
+func (e *ExperienceManager) GetExperienceToNextLevel() int {
```
Comment (T14Raptor) [inline server/entity/experience.go:39]:
```suggestion
func (e *ExperienceManager) ExperienceToNextLevel() int {
```

File: server/entity/experience.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	return e.GetExperienceForLevel(e.level)
+}
+
+// GetExperienceNettedToLevelUp get xp that you need to level up with reducing xp that you already have.
+func (e *ExperienceManager) GetExperienceNettedToLevelUp() int {
```
Comment (T14Raptor) [inline server/entity/experience.go:44]:
```suggestion
func (e *ExperienceManager) ExperienceNeededToLevelUp() int {
```

File: server/entity/experience.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	return e.GetExperienceToNextLevel() - e.ProgressToXp(e.progress, e.level)
+}
+
+// GetExperienceForLevel get the amount xp to level up in a specific level.
+func (e *ExperienceManager) GetExperienceForLevel(level int32) int {
```
Comment (T14Raptor) [inline server/entity/experience.go:49]:
```suggestion
func (e *ExperienceManager) ExperienceForLevel(level int32) int {
```

File: server/entity/experience.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+		return int(9*level - 158)
+	} else if level > 15 {
+		return int(5*level - 38)
+	} else {
+		return int(2*level + 7)
+	}
+}
+
+// GetExperienceForLevels calculate amount of xp for reach a level.
+func (e *ExperienceManager) GetExperienceForLevels(level int32) int {
```
Comment (T14Raptor) [inline server/entity/experience.go:61]:
```suggestion
func (e *ExperienceManager) ExperienceForLevels(level int32) int {
```

**Review by T14Raptor — COMMENTED** [review]
Left a few comments, I'll put a few more later

File: server/entity/experience.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
+package entity
+
+import "math"
+
+var maxLevel int32 = math.MaxInt32
+
+type ExperienceManager struct {
+	level    int32
```
[the remark this answers is not in the document]
Comment (Arcoz0308) [reply server/entity/experience.go:8]: what is a mutex ?

File: server/entity/experience.go [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+	progress float64
+	totalXP  int
```
[the remark this answers is not in the document]
Comment (Arcoz0308) [reply server/entity/experience.go:10]: that reduce calculs sow cpu usage, is not better ? can take many time if someone have 500 levels if you need to calculate all the levels

File: server/entity/experience.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+// GetExperienceForLevel get the amount xp to level up in a specific level.
+func (e *ExperienceManager) GetExperienceForLevel(level int32) int {
+	// I have make like MINET
+	if level > 30 {
+		return int(9*level - 158)
+	} else if level > 15 {
+		return int(5*level - 38)
+	} else {
```
Comment (Sandertv) [inline server/entity/experience.go:73]: Can remove this else.

File: server/entity/experience.go (hunk 7 above)
Comment (Sandertv) [reply server/entity/experience.go]: I think we should decide if we want to consistently go for `Experience` or for `XP`.

File: server/entity/experience.go (hunk 8 above)
Comment (Sandertv) [reply server/entity/experience.go]: Same as above.

**Review by Sandertv — COMMENTED** [review]
I think we'll also need to implement experience orbs if we want to go for a proper implementation.

**Comment by Arcoz0308** [comment]
i think i go close this pr, i'm not good enough in golang

File: server/entity/experience.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (T14Raptor) [reply server/entity/experience.go]: @Sandertv Opinion?

**Comment by Arcoz0308** [comment]
i don't know how finish experience orb so maybe someone other finish it or delete this PR

File: server/entity/experience.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (T14Raptor) [reply server/entity/experience.go]: This actually seemed to have over-complicate some of the code, I think it's best to keep track of the progress and total xp

[the remark this answers is not in the document]
Comment (Arcoz0308) [reply server/entity/experience.go]: Sandertv have say that he preffer to only save totalxp 🤷‍♂️

File: server/entity/experience.go [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+	nextLevel := true
+	for nextLevel {
+		xp := e.ExperienceForLevel(level)
+		if xp <= experience {
+			experience -= xp
+			level++
+		} else {
+			nextLevel = false
+		}
+	}
```
Comment (Sandertv) [inline server/entity/experience.go:35]: This code is duplicated everywhere. Might want to pull this out in its own method?

**Comment by T14Raptor** [comment]
Going to merge this to feature/experience and will continue it in the near future. Thanks!

**Pull request merged**